libraries/reliable_amqp_publisher/reliable_amqp_publisher.cpp

```diff
@@ -0,0 +1,247 @@
+
+#include <thread>
+#include <chrono>
+#include <future>
+#include <algorithm>
+
+#include <boost/asio/ip/tcp.hpp>
+#include <boost/filesystem.hpp>
+
+namespace eosio {
+
+struct reliable_amqp_publisher_callbacks {
+   virtual void amqp_ready() = 0;
+   virtual void amqp_error(const char* msg) = 0;
+};
+
+struct reliable_amqp_publisher_handler : AMQP::LibBoostAsioHandler {
+   reliable_amqp_publisher_handler(reliable_amqp_publisher_callbacks* callbacks, boost::asio::io_context& ctx) :
+      AMQP::LibBoostAsioHandler(ctx),
+      cbs(callbacks) {}
+
+   void onReady(AMQP::TcpConnection* c) override {
+      cbs->amqp_ready();
+   }
+
+   //disable heartbeat for now
+   uint16_t onNegotiate(AMQP::TcpConnection* connection, uint16_t interval) override {
+      return 0;
+   }
+
+   void onError(AMQP::TcpConnection*, const char* message) override {
+      cbs->amqp_error(message);
+   }
+
+   reliable_amqp_publisher_callbacks* cbs;
+};
+
+struct reliable_amqp_publisher_impl final : reliable_amqp_publisher_callbacks {
+   reliable_amqp_publisher_impl(const std::string url, const std::string exchange, const std::string routing_key, const boost::filesystem::path unconfirmed_path, const std::optional<std::string> message_id) :
+     data_file_path(unconfirmed_path), exchange(exchange), routing_key(routing_key), message_id(message_id) {
+      std::string host = "localhost", user = "guest", pass = "guest", path;
+      uint16_t port = 5672;
+      fc::url parsed_url(url);
+      FC_ASSERT(parsed_url.proto() == "amqp", "Only amqp:// URIs are supported for AMQP addresses");
+      if(parsed_url.host() && parsed_url.host()->size())
+         host = *parsed_url.host();
+      if(parsed_url.user())
+         user = *parsed_url.user();
+      if(parsed_url.pass())
+         pass = *parsed_url.pass();
+      if(parsed_url.port())
+         port = *parsed_url.port();
+      if(parsed_url.path())
+         path = parsed_url.path()->string();
+
+      std::stringstream hostss;
+      hostss << "amqp://" << user << ":" << pass << "@" << host << ":" << port << "/" << path;
+      amqp_address = std::make_unique<AMQP::Address>(hostss.str());
+
+      if(boost::filesystem::exists(data_file_path)) {
+         try {
+            fc::datastream<fc::cfile> file;
+            file.set_file_path(data_file_path);
+            file.open("rb");
+            fc::raw::unpack(file, message_deque);
+         } FC_RETHROW_EXCEPTIONS(error, "Failed to load previously unconfirmed AMQP messages from ${f}", ("f", (fc::path)data_file_path));
+      }
+      else {
+         boost::filesystem::ofstream o(data_file_path);
+         FC_ASSERT(o.good(), "Failed to create unconfirmed AMQP message file at ${f}", ("f", (fc::path)data_file_path));
+      }
+      boost::system::error_code ec;
+      boost::filesystem::remove(data_file_path, ec);
+
+      thread = std::thread([this]() {
+         try {
+            fc::set_os_thread_name("amqp");
+            retry_connection();
```

Comment:
Do we really want to wait 1 second before attempting to connect?

Reply:
yeah that has bothered me too; should probably be more aggressive on startup